### PR TITLE
CUMULUS-2129: Create collections migration from Dynamo -> RDS

### PR DESCRIPTION
**Pull request by markdboyd**
**Summary:** Summary of changes

Addresses [CUMULUS-2129: Create collections migration from Dynamo -> RDS](https://bugs.earthdata.nasa.gov/browse/CUMULUS-2129)

## Changes

* Detailed list or prose of changes
* ...

## PR Checklist

- [ ] Update CHANGELOG
- [ ] Unit tests
- [ ] Adhoc testing
- [ ] Integration tests

### CODE REVIEW

File: tf-modules/cumulus/data_migration.tf [hunk 1]
```diff
@@ -0,0 +1,19 @@
+module "data_migration_lambda" {
```
Comment (markdboyd) [inline tf-modules/cumulus/data_migration.tf:1]: You can probably ignore this file as it's likely to change pending the outcome of the conversation about how to run the data migration

**Comment by markdboyd** [comment]
The work to set up the deployment for the data migration and connect it to RDS is ongoing (and also blocked by #1826), but the actual code for running the data migration is reviewable

File: example/data-migration1-tf/main.tf [hunk 2]
```diff
@@ -0,0 +1,23 @@
+data "terraform_remote_state" "data_persistence" {
+  backend   = "s3"
+  config    = var.data_persistence_remote_state_config
+  workspace = terraform.workspace
+}
+
+module "data_migration1" {
```
Comment (markdboyd) [inline example/data-migration1-tf/main.tf:7]: If anyone has a better suggestion for the name of migration for collection data (and possibly later providers) that has to run **before** re-deploying Cumulus, I'm all ears

File: packages/db/src/connection.ts [hunk 3]
```diff
@@ -71,30 +105,7 @@ const buildKnexConfiguration = ({
...
-
-  const knexConfig = buildKnexConfiguration({
-    connectionConfig,
-    asyncStackTraces: env.KNEX_ASYNC_STACK_TRACES === 'true',
-    debug: env.KNEX_DEBUG === 'true',
-    migrationDir: env.migrationDir,
-    timeout,
-  });
-
+  const connectionConfig = await getConnectionConfig(env);
```
Comment (markdboyd) [inline packages/db/src/connection.ts:108]: @Jkovarik How do you feel about this refactor? It allows me to easily use the connection config to get the hostname here:

https://github.com/nasa/cumulus/pull/1828/files?file-filters%5B%5D=.ts#diff-88fa195f355a7e766958b02120bd5a2aR66

It also doesn't break any existing code

Comment (Jkovarik) [reply packages/db/src/connection.ts:108]: It's probably 'ok' - it's doubling the AWS SM cost of establishing a connection ($0.05 per 10,000 API calls), but only for provisioning that even cares about that info , that shouldn't cause heartburn. 

We should be careful we don't adopt this pattern on scaleable processes though, or if we do, memoizing the call or some other caching solution.

Comment (markdboyd) [reply packages/db/src/connection.ts:108]: How is it doubling the cost? The same amount of calls to retrieve the secret are made. I just broke down this function into two smaller functions so that in the provision database lambda I can inspect the connection config and then turn around and use it to build a Knex connection

File: lambdas/db-migration/package.json [hunk 4]
```diff
@@ -20,6 +22,7 @@
     "pg": "^8.3.0"
   },
   "devDependencies": {
-    "@types/pg": "^7.14.4"
+    "@types/pg": "^7.14.4",
+    "string-replace-loader": "^2.3.0"
```
Comment (yjpa7145) [inline lambdas/db-migration/package.json:26]: Since this is a dev dependency, it should live in the top-level `package.json`

File: example/data-migration1-tf/main.tf (hunk 2 above)
Comment (Jkovarik) [reply example/data-migration1-tf/main.tf:7]: cumulus-dynamo-to-rds-migration?

File: packages/db/src/connection.ts (hunk 3 above)
Comment (Jkovarik) [reply packages/db/src/connection.ts:108]: Right you are - when I read this yesterday I somehow thought `getKnexConfig` was re-making a call to SM.    Ignore the above, that's entirely 💯

File: packages/db/src/connection.ts [hunk 5]
```diff
@@ -42,6 +42,40 @@ const buildKnexConfiguration = ({
   return knexConfig;
 };
 
+export const getConnectionConfig = async (
```
Comment (Jkovarik) [inline packages/db/src/connection.ts:45]: We should probably docstring these as they're externally used now.

File: lambdas/data-migration1/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+import Logger from '@cumulus/logger';
+
+const {
+  Manager,
+} = require('@cumulus/api/models');
+const schemas = require('@cumulus/api/models/schemas');
+
+const logger = new Logger({ sender: '@cumulus/data-migration' });
+
+export interface HandlerEvent {
```
Comment (Jkovarik) [inline lambdas/data-migration1/src/index.ts:15]: Why is this needed over the default NodeJS.ProcessEnv?      This definition safely conforms to NodeJS.ProcessEnv without providing additional information.

File: lambdas/data-migration1/.gitignore [hunk 7]
```diff
@@ -0,0 +1,3 @@
+*.tfvars
+.terraform
```
Comment (yjpa7145) [inline lambdas/data-migration1/.gitignore:2]:
```suggestion
.terraform/
```

File: lambdas/data-migration1/src/index.ts (hunk 6 above)
Comment (Jkovarik) [reply lambdas/data-migration1/src/index.ts:15]: Particularly as `migrateCollections` takes NodeJS.ProcessEnv

File: lambdas/data-migration1/package.json [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    "@cumulus/api": "2.0.1",
+    "@cumulus/aws-client": "2.0.1",
+    "@cumulus/db": "2.0.1-alpha.1",
+    "@cumulus/logger": "2.0.1",
+    "knex": "0.21.5",
+    "pg": "^8.3.0"
+  },
+  "devDependencies": {
+    "@types/pg": "^7.14.4",
+    "aws-sdk": "^2.735.0",
```
Comment (yjpa7145) [inline lambdas/data-migration1/package.json:35]: These should be moved to the top-level `package.json`

File: lambdas/data-migration1/src/index.ts [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+  sampleFileName?: string
+  url_path?: string
+  ignoreFilesConfigForDiscovery?: boolean
+  meta?: object
+  tags?: string
+  created_at: Date
+  updated_at: Date
+}
+
+const getRequiredEnvVar = (name: string, env: NodeJS.ProcessEnv): string => {
```
Comment (Jkovarik) [inline lambdas/data-migration1/src/index.ts:39]: This method already exists in db/config - is there a reason for not using that here?

File: lambdas/db-migration/package.json [hunk 10]
```diff
@@ -10,7 +10,9 @@
   "scripts": {
     "clean": "rm -rf dist",
     "db:local:schema": "env PGPASSWORD=password pg_dump -h localhost -U postgres --schema-only --no-owner | egrep -v '^--' | cat -s | tee schema.sql",
-    "package": "npm run clean && npm run tsc && ../../node_modules/.bin/webpack && (cd dist/webpack && zip lambda.zip index.js)",
+    "webpack": "../../node_modules/.bin/webpack",
+    "build-lambda-zip": "(cd dist/webpack && cp -R ../lambda/migrations . && zip -r lambda.zip index.js migrations)",
```
Comment (yjpa7145) [inline lambdas/db-migration/package.json:14]:
```suggestion
    "build-lambda-zip": "(cd dist/webpack && cp -R ../lambda/migrations . && node ../../../bin/zip.js lambda.zip index.js migrations/*)",
```

File: lambdas/data-migration1/src/index.ts [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+
+export const migrateCollections = async (env: NodeJS.ProcessEnv, knex: Knex) => {
+  const collectionsTable = getRequiredEnvVar('CollectionsTable', env);
+
+  const searchQueue = new DynamoDbSearchQueue({
+    TableName: collectionsTable,
+  });
+  const createdRecordIds = [];
+
+  let record = await searchQueue.peek();
```
Comment (yjpa7145) [inline lambdas/data-migration1/src/index.ts:107]:
```suggestion
  let record = await searchQueue.shift();
```

File: lambdas/data-migration1/src/index.ts [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+  /* eslint-disable no-await-in-loop */
+  while (record) {
+    try {
+      const createdRecordId = await migrateCollectionRecord(record, knex);
+      if (createdRecordId) createdRecordIds.push(createdRecordId);
+    } catch (error) {
+      logger.error('Could not create collection record:', error);
+    }
+
+    await searchQueue.shift();
```
Comment (yjpa7145) [inline lambdas/data-migration1/src/index.ts:117]: These two lines can be reduced to:

```js
record = await searchQueue.shift();
```

File: lambdas/data-migration1/src/index.ts [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+    record = await searchQueue.peek();
+  }
+  /* eslint-enable no-await-in-loop */
+  logger.info(`successfully migrated ${createdRecordIds.length} collection records`);
+  return createdRecordIds;
```
Comment (yjpa7145) [inline lambdas/data-migration1/src/index.ts:122]: Since we aren't returning this return value, do we need to return it?

File: lambdas/data-migration1/src/index.ts [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+  return <number>cumulusId;
+};
+
+export const migrateCollections = async (env: NodeJS.ProcessEnv, knex: Knex) => {
+  const collectionsTable = getRequiredEnvVar('CollectionsTable', env);
+
+  const searchQueue = new DynamoDbSearchQueue({
+    TableName: collectionsTable,
+  });
+  const createdRecordIds = [];
```
Comment (yjpa7145) [inline lambdas/data-migration1/src/index.ts:105]: It's a little tricky that we are saving the IDs but not actually using them. Could this just be a counter instead?

File: lambdas/data-migration1/src/index.ts [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+
+  let record = await searchQueue.peek();
+  /* eslint-disable no-await-in-loop */
+  while (record) {
+    try {
+      const createdRecordId = await migrateCollectionRecord(record, knex);
+      if (createdRecordId) createdRecordIds.push(createdRecordId);
+    } catch (error) {
+      logger.error('Could not create collection record:', error);
```
Comment (yjpa7145) [inline lambdas/data-migration1/src/index.ts:114]: You may want to log the collection name and version here as well, to help with debugging.

File: lambdas/data-migration1/src/index.ts [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+    updated_at: new Date(dynamoRecord.updatedAt),
+  };
+
+  const [cumulusId] = await knex('collections')
+    .returning('cumulusId')
+    .insert(updatedRecord);
+  return <number>cumulusId;
+};
+
+export const migrateCollections = async (env: NodeJS.ProcessEnv, knex: Knex) => {
```
Comment (yjpa7145) [inline lambdas/data-migration1/src/index.ts:99]: Can you add the TS return type here?

File: lambdas/data-migration1/src/index.ts [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+ *   New record ID on success, false if record was skipped
+ */
+export const migrateCollectionRecord = async (
+  dynamoRecord: AWS.DynamoDB.DocumentClient.AttributeMap,
+  knex: Knex
+): Promise<number | false> => {
+  // Use schema to validate record before processing
+  Manager.recordIsValid(dynamoRecord, schemas.collection);
+
+  const [existingRecord] = await knex('collections')
```
Comment (yjpa7145) [inline lambdas/data-migration1/src/index.ts:63]:
```suggestion
  const [existingRecord] = await knex<RDSCollectionRecord>('collections')
```

File: lambdas/data-migration1/src/index.ts (hunk 14 above)
Comment (Jkovarik) [inline lambdas/data-migration1/src/index.ts:105]: It doesn't look like we use these in the return, if we need the length (ignoring my comment above about a report) could we just increment a counter?

File: lambdas/data-migration1/src/index.ts (hunk 16 above)
Comment (Jkovarik) [inline lambdas/data-migration1/src/index.ts:99]: This should probably have an explicit return type.

File: lambdas/data-migration1/src/index.ts [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+  logger.info(`successfully migrated ${createdRecordIds.length} collection records`);
+  return createdRecordIds;
+};
+
+export const handler = async (event: HandlerEvent): Promise<void> => {
+  const env = event?.env ?? process.env;
+  const knex = await connection.knex(env);
+
+  try {
+    await migrateCollections(env, knex);
```
Comment (Jkovarik) [inline lambdas/data-migration1/src/index.ts:130]: See above comment re: something giving a summary of complete/skipped/etc  to users

File: lambdas/data-migration1/src/index.ts (hunk 13 above)
Comment (Jkovarik) [inline lambdas/data-migration1/src/index.ts:122]: See above comment re: not using this value and/or possibly using it in a summary of the result

File: lambdas/data-migration1/variables.tf [hunk 19]
```diff
@@ -0,0 +1,38 @@
+variable "permissions_boundary_arn" {
```
Comment (Jkovarik) [inline lambdas/data-migration1/variables.tf:1]: Nit: alpha sort/label required variables

File: lambdas/data-migration1/src/index.ts [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+    url_path: dynamoRecord.url_path,
+    duplicateHandling: dynamoRecord.duplicateHandling,
+    granuleIdValidationRegex: dynamoRecord.granuleId,
+    granuleIdExtractionRegex: dynamoRecord.granuleIdExtraction,
+    // have to stringify on an array of values
+    files: JSON.stringify(dynamoRecord.files),
+    reportToEms: dynamoRecord.reportToEms,
+    sampleFileName: dynamoRecord.sampleFileName,
+    ignoreFilesConfigForDiscovery: dynamoRecord.ignoreFilesConfigForDiscovery,
+    meta: dynamoRecord.meta ? dynamoRecord.meta : undefined,
```
Comment (yjpa7145) [inline lambdas/data-migration1/src/index.ts:86]: Does this have to be stringified since it's json?

File: lambdas/db-migration/webpack.config.js [hunk 21]
```diff
@@ -39,7 +39,26 @@ module.exports = {
           },
         ],
       },
+      // NOTE: This is dark magic that prevents Knex from failing when
```
Comment (Jkovarik) [inline lambdas/db-migration/webpack.config.js:42]: Please add a warning in the README for this lambda re: this, as an update to knex could cause all manner of grief.

File: packages/aws-client/src/DynamoDbSearchQueue.ts [hunk 22]
```diff
@@ -1,3 +1,4 @@
+import * as AWS from 'aws-sdk';
```
Comment (Jkovarik) [inline packages/aws-client/src/DynamoDbSearchQueue.ts:1]: I'm sure I'm just missing something obvious, but why this change?

Comment (Jkovarik) [reply packages/aws-client/src/DynamoDbSearchQueue.ts:1]: Nevermind, ignore the above.

File: lambdas/data-migration1/src/index.ts (hunk 6 above)
Comment (markdboyd) [reply lambdas/data-migration1/src/index.ts:15]: There is probably no reason. This was copypasta. However, is it not helpful to specify that the handler event may contain a key `env` and that key should conform to NodeJS.ProcessEnv

File: lambdas/data-migration1/src/index.ts (hunk 9 above)
Comment (markdboyd) [reply lambdas/data-migration1/src/index.ts:39]: I don't totally love that a general utility for getting required env vars lives in the `db` package.  I hate dumping things in `common` and creating that dependency though

Comment (yjpa7145) [reply lambdas/data-migration1/src/index.ts:39]: `common` or "its own package" are the only options that seem to make sense. I don't have a problem with this living in `common`, especially as we've been pruning `common` down to utilities that make more sense.

File: lambdas/data-migration1/src/index.ts (hunk 6 above)
Comment (yjpa7145) [reply lambdas/data-migration1/src/index.ts:15]: This seems fine, it's saying that an `env` can be passed in through the event, and using `NodeJS.ProcessEnv` seems the right type.

File: lambdas/data-migration1/src/index.ts (hunk 13 above)
Comment (markdboyd) [reply lambdas/data-migration1/src/index.ts:122]: I'm using it in the tests, which isn't necessarily a good answer. But it can easily be refactored out

File: lambdas/data-migration1/src/index.ts (hunk 12 above)
Comment (markdboyd) [reply lambdas/data-migration1/src/index.ts:117]: That doesn't seem to work because `shift` can possibly return `undefined`, causing a type mismatch

File: lambdas/data-migration1/src/index.ts (hunk 17 above)
Comment (markdboyd) [reply lambdas/data-migration1/src/index.ts:63]: Not that I'm against it, but what does this add for us? Also the interface `RDSCollectionRecord` doesn't specify `cumulusId` as a property since it is used for the data that is passed before a record is saved, but what is returned here (if anything is returned) should have one.

File: lambdas/data-migration1/src/index.ts (hunk 20 above)
Comment (markdboyd) [reply lambdas/data-migration1/src/index.ts:86]: Only arrays of data have to be stringified, not objects

File: example/data-migration1-tf/main.tf (hunk 2 above)
Comment (markdboyd) [reply example/data-migration1-tf/main.tf:7]: It's not going to be the only dynamo-to-rds migration though. As I understood, the plan was to have this first migration for data that granules reference (e.g. collections, providers), which has to run **before** deploying the version of Cumulus that does parallel writes. Then there would be a second data migration that is included with the Cumulus deployment to migrate existing granules/files data. So I don't know how to distinguish between them beyond calling this migration number 1 and the second migration number 2.

File: lambdas/data-migration1/src/index.ts (hunk 15 above)
Comment (markdboyd) [reply lambdas/data-migration1/src/index.ts]: 66cd7c988c6d490117f9b7aab0aabc29fab9c578

File: lambdas/data-migration1/src/index.ts [hunk 23]
```diff
@@ -0,0 +1,134 @@
...
+ *   Source record from DynamoDB
+ * @param {Knex} knex - Knex client for writing to RDS database
+ * @returns {number|false}
+ *   New record ID on success, false if record was skipped
+ */
+export const migrateCollectionRecord = async (
+  dynamoRecord: AWS.DynamoDB.DocumentClient.AttributeMap,
+  knex: Knex
+): Promise<number | false> => {
+  // Use schema to validate record before processing
```
[the remark this answers is not in the document]
Comment (markdboyd) [reply lambdas/data-migration1/src/index.ts]: d27c3a6bdb682102f2f48163047c802fb15868d2

File: lambdas/data-migration1/src/index.ts [hunk 24]
```diff
@@ -0,0 +1,134 @@
...
+  Manager.recordIsValid(dynamoRecord, schemas.collection);
+
+  const [existingRecord] = await knex('collections')
+    .where('name', dynamoRecord.name)
+    .where('version', dynamoRecord.version);
+  // Skip record if it was already migrated.
+  if (existingRecord) {
+    logger.info(`Collection name ${dynamoRecord.name}, version ${dynamoRecord.version} was already migrated, skipping`);
```
[the remark this answers is not in the document]
Comment (markdboyd) [reply lambdas/data-migration1/src/index.ts:61]: How about d27c3a6bdb682102f2f48163047c802fb15868d2?

File: lambdas/db-migration/webpack.config.js (hunk 21 above)
Comment (markdboyd) [reply lambdas/db-migration/webpack.config.js:42]: 14f5a2fb628a950e221d1116c3f2e861f9fbf9db

File: lambdas/data-migration1/src/index.ts (hunk 16 above)
Comment (markdboyd) [reply lambdas/data-migration1/src/index.ts]: d27c3a6bdb682102f2f48163047c802fb15868d2

File: lambdas/data-migration1/src/index.ts (hunk 9 above)
Comment (markdboyd) [reply lambdas/data-migration1/src/index.ts]: 789f1e2f1c4a9884c365a6f3ff28a9a2064b1b0d

File: lambdas/data-migration1/src/index.ts (hunk 14 above)
Comment (markdboyd) [reply lambdas/data-migration1/src/index.ts]: d27c3a6bdb682102f2f48163047c802fb15868d2

File: lambdas/data-migration1/src/index.ts (hunk 12 above)
Comment (yjpa7145) [reply lambdas/data-migration1/src/index.ts:116]: Yeah, that's a problem with the way that we are handling typing for the `DynamoDbSearchQueue`. I won't hold up this PR over it, but I am going to issue another PR that fixes that.

File: lambdas/data-migration1/src/index.ts (hunk 17 above)
Comment (yjpa7145) [reply lambdas/data-migration1/src/index.ts:56]: In this particular case I guess nothing, since we aren't actually doing anything with `existingRecord`.

**Comment by yjpa7145** [comment]
Just a couple small changes requested for the `common/env` module. Other than that, my concerns have been addressed.

File: lambdas/data-migration1/src/index.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (Jkovarik) [reply lambdas/data-migration1/src/index.ts]: It's probably^H^H^H^H^H^H^H definitely a nit, and perhaps a bigger concern for files/granules, but the concern here was more about how we don't actually trap the error set and dump it to CW logs, so users will have to text munge that rather than having a concise list of what failed.    Success/Total might be good as well.

File: example/data-migration1-tf/main.tf (hunk 2 above)
Comment (Jkovarik) [reply example/data-migration1-tf/main.tf:7]: Right, sorry, I mean dynamo-to-rds-migration{n}.   It's not a great suggestion, frankly, so we can leave that be.

File: lambdas/data-migration1/src/index.ts (hunk 24 above)
[the remark this answers is not in the document]
Comment (markdboyd) [reply lambdas/data-migration1/src/index.ts:61]: Re:

>It's probably^H^H^H^H^H^H^H definitely a nit, and perhaps a bigger concern for files/granules, but the concern here was more about how we don't actually trap the error set and dump it to CW logs, so users will have to text munge that rather than having a concise list of what failed. Success/Total might be good as well.

I can definitely update the summary to show the total number of records migrated out of the total number in Dynamo

[the remark this answers is not in the document]
Comment (Jkovarik) [reply lambdas/data-migration1/src/index.ts:61]: (Thanks github! ) 

Given this is user facing, and they're going to have failed migrations at some point.... I think we probably should tell them at least the totals for success/fail/skip with a total dynamo count.

File: lambdas/data-migration1/src/index.ts (hunk 9 above)
Comment (Jkovarik) [reply lambdas/data-migration1/src/index.ts]: 👍  that seems fine, or at least resolves the concern.

**Comment by Jkovarik** [comment]
The updates look great - I think my only outstanding concern is the user facing output from the migration itself: https://github.com/nasa/cumulus/pull/1828#discussion_r482236464

**Review by yjpa7145 — APPROVED** [review]
I made the changes that I requested in ba96e5a774ef133550e0c64ed73823a74cd08b8d, merged master, and am marking this as "Approved".

**Comment by markdboyd** [comment]
@Jkovarik how about this? 07b1017211b8fe893b831db1de7a842ec3852315

File: lambdas/data-migration1/src/index.ts [hunk 25]
```diff
@@ -58,8 +69,7 @@ export const migrateCollectionRecord = async (
     .where('version', dynamoRecord.version);
   // Skip record if it was already migrated.
```
Comment (Jkovarik) [inline lambdas/data-migration1/src/index.ts:70]: Nit:  In the context of the function it's not longer skipping.  Take or leave, shouldn't hold up PR.

**Review by Jkovarik — APPROVED** [review]
> @Jkovarik how about this? [07b1017](https://github.com/nasa/cumulus/commit/07b1017211b8fe893b831db1de7a842ec3852315)

Looks good, thank you!    All of my concerns have been addressed, nice work!
